### ISSUE TITLE
Number of Builder sections on a page is limited by the web server's max_input_vars setting

### ISSUE THREAD

**Issue by coreymckrill** [thethemefoundry/make#264]
Each Builder section added to a page adds a certain number of inputs to the post form that gets submitted when the post is published or updated. Sections with sub items like the Gallery and the Banner will add more inputs for each additional item. Each of these inputs counts as a separate input variable when the form posts the data to the server, even if the post data is compiled as an array.

Problems arise, then, when there are more input variables than the `max_input_vars` limit that has been set on a server's PHP configuration. Shared servers tend to have a lower limit, and they also often put restrictions on changing PHP's configuration variables at runtime. Sections that push the number of input variables over the limit won't get saved when the page is published or updated.

The workaround to this, on a case-by-case basis, is to have the user change `max_input_vars` to a high number, like `5000`.

So the question is, **is there a different way to handle all of this page builder data so that we can avoid the PHP input variables limit altogether?**

---

**Original description:**

One user reported that any sections added after 30 were not retained after hitting the page's Update button.

https://thethemefoundry.com/questions/topic/error-with-update-page/

I was able to get to 35 before this happened, so it sounds like this is probably due to some bottleneck in the Builder architecture that is somehow affected by the server configuration.

**Comment by coreymckrill** [comment]
https://thethemefoundry.com/questions/topic/upgrade-to-1-6-1-prevents-updates-to-a-page-being-persisted/

**Comment by coreymckrill** [comment]
Here's an article that contains a potential workaround and possibly sheds some light on the cause of the issue:

http://www.wpbeginner.com/wp-tutorials/how-to-fix-the-custom-menu-items-limit-in-wordpress/

**Comment by coreymckrill** [comment]
From the second support topic above, adding a **php.ini** file, as suggested in the WP Beginner link, seems to resolve the issue with this line:

```
max_input_vars = 5000
```

**Comment by coreymckrill** [comment]
Another successful mitigation:
https://thethemefoundry.com/questions/topic/is-there-a-limit-to-the-number-of-rows-or-sections-on-a-builder-page/

**Comment by tollmanz** [comment]
@coreymckrill I ran into this issue last week. `max_input_vars` is definitely the culprit.

**Comment by coreymckrill** [comment]
@tollmanz Aye, it's a tricky one to fix, too. I suspect we will need to do a significant refactor in order to mitigate that limitation...

**Comment by coreymckrill** [comment]
@tollmanz any thoughts on how to approach a solution to this?

One idea I have is to consolidate all the inputs for a particular section into a JSON string, so that only one input per section gets submitted.

**Comment by tollmanz** [comment]
@coreymckrill That could work; however, when I experienced this bug, I saw that anything past the limit is simply dropped. Theoretically, if the ordering of the fields was such that the JSON string input was still past the limit, you'd still lose data.

I was thinking that you could have a single input store data and ensure it is the top-most field in the form. Whenever data is changed in one of the fields, this "master" field could be updated. While this would work, it could be a major pain as you have to monitor every input and update as it changes.

What would you think about using `ini_set` to change `max_input_vars`? I'm not sure if it would even work as I do not know when PHP trims the variables. Perhaps this is only configurable in the **php.ini** file. It's worth testing though.

**Comment by coreymckrill** [comment]
> What would you think about using `ini_set` to change `max_input_vars`?

I suspect that not all hosts allow modification of the PHP config variables [in this way](http://stackoverflow.com/questions/2095641/why-is-phps-ini-set-function-switched-off-on-shared-hosts), or at all, in some cases, so it wouldn't be a reliable solution, and it would be really hard to test.

> Theoretically, if the ordering of the fields was such that the JSON string input was still past the limit, you'd still lose data.

True. I've been thinking about how to get around that. I have read that inputs without a `name` attribute don't actually get submitted by the form with the rest of the POST data. I'm not 100% confident yet that this is true across all browsers, but it would definitely help alleviate the issue. If the name attributes were removed from all the individual setting inputs, and only used on hidden fields containing JSON strings, the number of actual input vars submitted would be reduced dramatically.

> While this would work, it could be a major pain as you have to monitor every input and update as it changes.

Yes, this would require redesigning the Backbone View objects to listen for changes on inputs and selects and syncing the new values to correlated Model objects. Then a JSON string with all the setting values could be extracted from the Model and fed into the hidden field. Seems convoluted, I know.

But another question is, would it be possible to do this without relying on submitting the data through a form at all? I've toyed with the notion that setting values could be saved asynchronously to a post revision or something as they are updated. I just haven't learned enough about the revision system to know if/how the data could then be copied to the actual post entry in the database when the post is published/updated.
